sos/report/plugins/kubernetes.py

```diff
@@ -123,57 +149,71 @@ def setup(self):
+        kns = self.collect_cmd_output(f'{self.kube_cmd} get namespaces')
         # namespace is the 1st word on line, until the line has spaces only
         kn_output = kns['output'].splitlines()[1:]
         knsps = [n.split()[0] for n in kn_output if n and len(n.split())]
 
         for nspace in knsps:
             knsp = '--namespace=%s' % nspace
             if self.get_option('all'):
-                k_cmd = '%s %s %s' % (self.kube_cmd, "get -o json", knsp)
-
-                self.add_cmd_output('%s events' % k_cmd)
+                k_cmd = f'{self.kube_cmd} get -o json {knsp}'
 
                 for res in self.resources:
-                    self.add_cmd_output('%s %s' % (k_cmd, res), subdir=res)
+                    self.add_cmd_output(f'{k_cmd} {res}', subdir=res)
 
             if self.get_option('describe'):
                 # need to drop json formatting for this
-                k_cmd = '%s %s' % (self.kube_cmd, knsp)
+                k_cmd = f'{self.kube_cmd} {knsp}'
                 for res in self.resources:
-                    ret = self.exec_cmd('%s get %s' % (k_cmd, res))
+                    ret = self.exec_cmd(f'{k_cmd} get {res}')
                     if ret['status'] == 0:
                         k_list = [k.split()[0] for k in
                                   ret['output'].splitlines()[1:]]
                         for item in k_list:
-                            k_cmd = '%s %s' % (self.kube_cmd, knsp)
+                            k_cmd = f'{self.kube_cmd} {knsp}'
                             self.add_cmd_output(
-                                '%s describe %s %s' % (k_cmd, res, item),
+                                f'{k_cmd} describe {res} {item}',
                                 subdir=res
                             )
 
             if self.get_option('podlogs'):
-                k_cmd = '%s %s' % (self.kube_cmd, knsp)
-                ret = self.exec_cmd('%s get pods' % k_cmd)
+                k_cmd = f'{self.kube_cmd} get -o json {knsp}'
+                ret = self.exec_cmd(f'{k_cmd} pods')
                 if ret['status'] == 0:
-                    pods = [p.split()[0] for p in
-                            ret['output'].splitlines()[1:]]
+                    pods = json.loads(ret['output'])
                     # allow shell-style regex
                     reg = (translate(self.get_option('podlogs-filter')) if
                            self.get_option('podlogs-filter') else None)
-                    for pod in pods:
-                        if reg and not re.match(reg, pod):
+                    for pod in pods["items"]:
+                        if reg and not re.match(reg, pod["metadata"]["name"]):
                             continue
-                        self.add_cmd_output('%s logs %s' % (k_cmd, pod),
-                                            subdir='pods')
+                        _subdir = f'pods/{pod["metadata"]["namespace"]}'
```

Comment:
Just a thought here, should we make the subdir path `namespaces/{namespace}/pods`? So that other namespace-specific items can be grouped under the namespace dir and then have another subdir for the resource type?

We'd want to also update the `all` collection for resources to this new subdir layout here, if we do this.

Reply:
I added the namespace subdir for pod logs since there will be too many files in pods/ if the k8s cluster has a lot of pods. It would be good to have subdirs so they can be separated.
At the moment the structure is grouped by resource type.
We can certainly change to grouped by namespace, just like what `kubectl cluster-info dump` do.
I can make the change if another developer gives another +1.

Reply:
@shundezhang i think it makes sense do do what @TurboTurtle suggested

Reply:
I have made the change to dir structure.
Now everything that is queried from k8s APIs, e.g. using kubectl command, is placed under cluster-info/
At the same level there are outputs from systemctl/journalctl commands, and maybe other system commands too.
Under cluster-info/ subdir, there are outputs for global resource and subdirs for each namespace.
Inside each namespace dir, a dir for each resource type holds outputs for kubectl describe outputs, and also there is a dir for pod logs. Json output files are placed under this namespace dir since there is only one file for each resource.
Examples:
1. kubernetes.all=on kubernetes.podlogs=on kubernetes.describe=on
https://paste.ubuntu.com/p/DCrSK92TYF/
2. kubernetes.all=off kubernetes.podlogs=off kubernetes.describe=off
https://paste.ubuntu.com/p/24fzFQ3kXQ/
3. kubernetes.describe=on kubernetes.podlogs=on
https://paste.ubuntu.com/p/mjS66wqT2W/
4. kubernetes.all=on kubernetes.podlogs=on
https://paste.ubuntu.com/p/C95sBQFjYJ/

Reply:
We should place the namespaces under a dir that makes it obvious what the subdirs are. Given that there is other content under `cluster-info` such as the global resources it's not immediately obvious what, for example, the `default` directory actually _is_.

The OpenShift plugin puts these under `namespaces`, e.g. the dir layout would be `sos_commands/openshift/namespaces/$namespace` - I think we should mimic that here. The global resources being under `cluster-info` is fine though, and it makes more sense than putting that stuff in the top-level plugin directory.

Reply:
Hi @TurboTurtle thank you for the feedback. The structure here is similar to the output of command `kubectl cluster-info dump --output-directory='./cluster-info' -A`. An example can be found here: https://paste.ubuntu.com/p/xXHRc9zfwP/
It doesn't have too many layers of dirs.
This structure should be easy to understand for a k8s user as it is the default output of kubectl command.
To make it more obvious, I can add `namespace_` to the beginning of those subdirs so they look like:
```
namespace_default\...
namespace_kube-system\...
```
What do you think?
Also could you please share a full output from openshift plugin to compare?

Reply:
Oh, ok. I didn't realize that was how it was dropped by default. In that case, no issue from me with the current state of how the output gets written to the archive.